### ISSUE TITLE
Is mirroring working with MXNet 1.5.1 Gluon ?

### ISSUE THREAD

**Issue by Wallart** [apache/incubator-mxnet#16754]
Hello everyone,
I am currently implementing a Sparse/LogSparse transformer.
I would like some layers results to be re-computed at backward pass to save memory.

I tried the MXNET_BACKWARD_DO_MIRROR env variable.
My network is hybridizable so I also tried to call _set_attr(force_mirroring='True') on the generated symbols.

But, it doesn't seems to have any effect on memory consumption.

Is mirroring supposed to work on Gluon API under MXNet 1.5.1 ?

**Comment by samskalicky** [comment]
@zachgk assign @szha

**Comment by szha** [comment]
mirror option isn't added to Gluon yet.

**Comment by Wallart** [comment]
Can you confirm that mirroring is still working on MXNet 1.5.1 using the symbolic API ?
As a placeholder, I will re-write my network using symbols.

**Comment by szha** [comment]
Yes, mirror should be working with the symbolic API/graph executor. There shouldn't be a need for rewriting, since HybridBlock outputs a symbol when you feed it with one.

**Comment by Wallart** [comment]
Ok, I will give it a try and let you know if it works

**Comment by Wallart** [comment]
Could you provide a basic example ? 
I'm struggling to find valid documentation on the use of HybridBlocks with a graph executor

**Comment by leezu** [comment]
`output_symbols = net(*input_symbols)` where `net` is your HybridBlock.
Just use the `output_symbols` and ignore `net`. You only use `net` to obtain the symbol.

**Comment by Wallart** [comment]
To test mirroring, I've built a quick example, however after the first iteration the loss is always nan. I cannot push investigation further.

I am missing something but I can't find what it is. I suppose that the Transformer LayerNorm / Dropout are not working in training mode, but using an autograd.train_mode() when I'm extracting the network symbols does not solve the issue.
https://gist.github.com/Wallart/c807af4fdd520a172f41b7e17badc296

Finding good examples on Module API  with the 404 errors on MXNet documentation feels like archeology :D

**Comment by leezu** [comment]
Can you post a minimal example to reproduce the problem?

**Comment by Wallart** [comment]
Yes of course I will create a repo.

**Comment by Wallart** [comment]
https://github.com/Wallart/transformer-ts-forecasting
For execution :
./main.py train synth ~/fakePath 24 -e 400 -b 64 --num-layers 2 --dim-model 32 --dim-ff 128

By default it will start a training using Gluon API. To switch on Module API, you have to paste the following lines

in main.py line 131
```
        #trainer = TransformerTrainer(args, net, ctx)
        trainer = ModuleTrainer(args, net, ctx)
```
in transformer.py line 38
```
        #if autograd.is_training():
        #    return logits, attn_w_1, attn_w_2
```

**Comment by Wallart** [comment]
My bad  a MXNET_BACKWARD_DO_MIRROR=1 was dragging around PyCharm run conf.

**Comment by leezu** [comment]
@Wallart so you mean everything is working as expected? Then I'll close the issue. But please reopen if I misunderstood.

**Comment by Wallart** [comment]
I'm still not sure how to use _set_attr(force_mirroring='True'). I can't see any impact on memory consumption.
Can I set the attribute in the HybridBlock hybrid_forward ? Or do I have to wait to get the full computation graph ?

**Comment by leezu** [comment]
I think you can set it on the symbol which you obtained via `output_symbols = net(*input_symbols)`.

**Comment by Wallart** [comment]
I confirm that mirroring is still working.
We can apply _set_attr(force_mirroring='True') on HybridBlocks and then work with the Module API
